shadowsocks/stream.go

```diff
@@ -61,25 +61,28 @@ func (sw *shadowsocksWriter) init() (err error) {
 		if _, err := io.ReadFull(rand.Reader, salt); err != nil {
 			return fmt.Errorf("failed to generate salt: %v", err)
 		}
-		_, err := sw.writer.Write(salt)
-		if err != nil {
-			return fmt.Errorf("failed to write salt: %v", err)
-		}
 		sw.aead, err = sw.ssCipher.Encrypter(salt)
 		if err != nil {
 			return fmt.Errorf("failed to create AEAD: %v", err)
 		}
 		sw.counter = make([]byte, sw.aead.NonceSize())
-		sw.buf = make([]byte, 2+sw.aead.Overhead()+payloadSizeMask+sw.aead.Overhead())
+		// The maximum length message is the salt (first message only), length, length tag,
+		// payload, and payload tag.
+		sizeBufSize := 2 + sw.aead.Overhead()
+		maxPayloadBufSize := payloadSizeMask + sw.aead.Overhead()
+		sw.buf = make([]byte, len(salt)+sizeBufSize+maxPayloadBufSize)
```

Comment:
Move this earlier and delete the `make` call for the salt. You might as well write the salt directly to this buffer.

Reply:
I tried that initially, but it doesn't work because we need to allocate `salt` so that we can construct `sw.aead`, and we need `sw.aead` to know what the overhead is.  We could solve this by setting a max overhead, but currently we don't have one defined.

Reply:
I see. Ugh, it would be helpful if we had descriptors for all ciphers available.